modules/porous_flow/src/bcs/PorousFlowEnthalpySink.C

```diff
@@ -0,0 +1,89 @@
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PorousFlowEnthalpySink.h"
+#include "Function.h"
+#include "PorousFlowDictator.h"
+#include "SinglePhaseFluidProperties.h"
+#include "MooseVariable.h"
+
+#include "libmesh/quadrature.h"
+
+#include <iostream>
+
+registerMooseObject("PorousFlowApp", PorousFlowEnthalpySink);
+
+InputParameters
+PorousFlowEnthalpySink::validParams()
+{
+  InputParameters params = IntegratedBC::validParams();
+  params.addRequiredParam<UserObjectName>(
+      "PorousFlowDictator", "The UserObject that holds the list of PorousFlow variable names");
+  params.addParam<unsigned int>(
+      "fluid_phase", "If supplied, then this BC will potentially be a function of fluid pressure.");
+  params.addRequiredParam<FunctionName>(
+      "flux_function",
+      "The mass flux.  The flux is OUT of the medium: hence positive values of "
+      "this function means this BC will act as a SINK, while negative values "
+      "indicate this flux will be a SOURCE.  The functional form is useful "
+      "for spatially or temporally varying sinks.");
+  params.addRequiredParam<UserObjectName>("fp", "The name of the user object for fluid properties");
+  params.addRequiredCoupledVar("pressure", "Pressure");
+  params.addRequiredParam<Real>("T_in", "Specified inlet temperature");
+  params.addClassDescription("Applies a flux sink to a boundary.");
+
+  return params;
+}
+
+PorousFlowEnthalpySink::PorousFlowEnthalpySink(const InputParameters & parameters)
+  : IntegratedBC(parameters),
+    _dictator(getUserObject<PorousFlowDictator>("PorousFlowDictator")),
+    _involves_fluid(isParamValid("fluid_phase")),
+    _ph(_involves_fluid ? getParam<unsigned int>("fluid_phase") : 0),
+    _m_func(getFunction("flux_function")),
+    _pp(getMaterialProperty<std::vector<Real>>("PorousFlow_porepressure_nodal")),
+    _dpp_dvar(
+        getMaterialProperty<std::vector<std::vector<Real>>>("dPorousFlow_porepressure_nodal_dvar")),
+    _T_in(getParam<Real>("T_in")),
+    _fp(getUserObject<SinglePhaseFluidProperties>("fp"))
+{
+}
+
+Real
+PorousFlowEnthalpySink::computeQpResidual()
+{
+  Real h = _fp.h_from_p_T(_pp[_qp][_ph], _T_in);
+  return _test[_i][_qp] * _m_func.value(_t, _q_point[_qp]) * h;
+}
+
+Real
+PorousFlowEnthalpySink::computeQpJacobian()
+{
+  return jac(_var.number());
+}
+
+Real
+PorousFlowEnthalpySink::computeQpOffDiagJacobian(unsigned int jvar)
+{
+  return jac(jvar);
+}
+
+Real
+PorousFlowEnthalpySink::jac(unsigned int jvar) const
+{
+  if (_dictator.notPorousFlowVariable(jvar))
+    return 0.0;
+
+  const unsigned int pvar = _dictator.porousFlowVariableNum(jvar);
+
+  Real h, dh_dpp, dh_dT;
+  _fp.h_from_p_T(_pp[_qp][_ph], _T_in, h, dh_dpp, dh_dT);
+  Real jac = dh_dpp * _dpp_dvar[_i][_ph][pvar];
+  return _test[_i][_qp] * _phi[_j][_qp] * jac;
```

Comment:
Again, think about upwinding.  If you use the upwinded version, these lines should be
```
if (_i != _j)
  return 0.0;
_fp.h_from_p_T(_pp[_i][_ph], _T_in, h, dh_dpp, dh_dT);
const Real jac = dh_dpp * _dpp_dvar[_i][_ph][pvar];
return _test[_i][_qp] * jac * _m_func.value(_t, _q_point[_qp])
```
If you don't want to upwind, then you should use `_dpp_dvar[_qp][_ph][pvar]` and include the `_m_func.value(_t, _q_point[_qp])`  (and pull in the "_qp" version of the derivative, as mentioned above)